markdown/util.py

```diff
@@ -378,9 +379,21 @@ def _sort(self):
 
         This method is called internally and should never be explicitly called.
         """
-        if not self._is_sorted:
-            self._priority.sort(key=lambda item: item.priority, reverse=True)
-            self._is_sorted = True
+        self._priority.sort(key=lambda item: item.priority, reverse=True)
+        self._is_sorted = True
+
+    @property
+    def _is_sorted(self):
+        return self._sort is _noop
+
+    @_is_sorted.setter
+    def _is_sorted(self, is_sorted):
+        if is_sorted:
+            # Subsequent calls to _sort() will do nothing.
+            self._sort = _noop
+        else:
+            # Restore the original sort function.
+            self.__dict__.pop('_sort', None)
```

Comment:
While I understand what is going on here, this feels a little too much like "magic." The next time I look at this code, I may not remember why we are jumping through these hoops. And that means that next time this section of code has any work done, it's likely this would just get ripped out. As a reminder, performance has never been a top priority for us.

Reply:
Yes, in this shape with the 2nd commit it is quite magic. Could you please check out how it looks after just the **first commit** in this PR, and whether we could get *that* into an acceptable shape? I think with enough comments it can be fine.

I can sympathize with the general sentiment. But note that I already filtered out some changes and left only the juiciest ones. And if you consider how many times this code runs on a global scale, it becomes easier to justify.

Reply:
So, the options are to either:

1. Call a random builtin C function (you happened to pick `bool`), or
2. Call a Python function which executes one line (`if not self._is_sorted` which evaluates `False`).

The only reason (1) is even an option is because we always ignore the return value. Otherwise, we would need to stick with the Python function.

This seems to really be slitting hairs. I just don't care enough to bother with it and the potential maintenance headache in the future fare outweighs any potential gains. The only reason I haven't closed the PR completely is because the other change (in treeprocessors.py) is a reasonable change.

Reply:
Makes sense.
I'm also now seeing that running under the profiler seems to give Python functions a disadvantage, so the effect of the `_sort` change is even smaller. The `__len__` one is still the bigger one and is noticeable, so let's keep it.